src/server/master_server/metadata_manager.cc

```diff
@@ -9,10 +10,33 @@ namespace server {
 
 MetadataManager::MetadataManager() {
   lock_manager_ = LockManager::GetInstance();
+  // Set the submap number and initialize locks and each submap
+  num_of_submap_ = std::max(std::thread ::hardware_concurrency(),
+                       (unsigned int)1);
+  file_metadata_lock_ = std::vector<absl::Mutex*>(num_of_submap_, 
+                            new absl::Mutex());
+  file_metadata_ = std::vector<absl::flat_hash_map<std::string,
+                       std::shared_ptr<protos::FileMetadata>>>(num_of_submap_);
+  chunk_metadata_lock_ = std::vector<absl::Mutex*>(num_of_submap_,
+                            new absl::Mutex());
+  chunk_metadata_ = std::vector<absl::flat_hash_map<std::string,
+                        protos::FileChunkMetadata>>(num_of_submap_);
+}
+
+inline size_t MetadataManager::submap_id(const std::string& filename) const {
+  // Compute the hash of given string, and mod the number of submaps
+  return std::hash<std::string>{}(filename) % num_of_submap_;
```

Comment:
We are duplicating code here from lock manager. Seems like we need a class for ThreadSafeFlatHashMap. Will enable us to easily reuse for other components. Including chunkserver manager. Don't have to have a full implementation for now. Will also add to it, once I start using it. Thanks.

Reply:
BTW, please what's the reason for abandoning the parallel hash map? I already started using it

Reply:
Ideally yes, let me see if I can generalize it (has to play with template stuff a bit).

Reply:
Good question about the parallel hash map, you could see some points I raised in our slack channel under idea section if I remember correctly. 

Long story short, I have seen inconsistent behavior when non-trivial types are given to the parallel hashmap. I see that an insertion takes place, but later .at() returns null, and there is no deletion. Regardless, without having explicit control of the locks for submaps is a pain for our usecase, because we want something like an atomic operation that "give me the value or tell me there is an error". Consider how you would do this with a parallel hash map, you'd have to first check and then access, but that leaves a time-of-check-time-of-use issue. The library claims to give "if_contains" + an anonymous call back, but I think it is not well tested as I ran into compilation issues using it. In any case, story gets complicated once you start using it with care. 

Having the explicit control of locks, as you see here appears to resolve the strange behavior I observed in the stress unit test.

Reply:
Sounds good. Thanks for the explanation. Sorry I missed the discussion.